File: bankflow_web/contracts.py
"""Stable, deliberately small DTO contracts exposed to the Web frontend."""

from __future__ import annotations

from dataclasses import asdict, dataclass, field
from typing import Any


ERROR_MESSAGES = {
    "NO_CASE": "尚未打开标准结果",
    "FILE_NOT_FOUND": "所选标准结果不存在",
    "INVALID_JSON": "所选文件不是有效的 JSON",
    "SCHEMA_INCOMPATIBLE": "该结果版本不受当前程序支持",
    "INVALID_ARGUMENT": "请求参数无效",
    "STALE_CASE": "案件已切换，本次请求已过期",
    "TRANSACTION_NOT_FOUND": "未找到指定交易",
    "EVIDENCE_UNAVAILABLE": "该交易的证据不可用",
    "FRONTEND_NOT_READY": "前端尚未准备完成",
    "CASE_DIRECTORY_NOT_FOUND": "所选案件目录不存在",
    "CASE_DIRECTORY_READ_FAILED": "无法读取所选案件目录",
    "CASE_HANDLE_INVALID": "案件选择已失效，请重新选择",
    "NO_SUPPORTED_SOURCES": "目录中没有可处理的 PDF/Excel 流水文件",
    "ANALYSIS_ALREADY_RUNNING": "已有分析任务正在运行",
    "ANALYSIS_TASK_NOT_FOUND": "分析任务不存在或已经失效",
    "ANALYSIS_STILL_RUNNING": "分析仍在进行，暂时不能关闭任务",
    "ANALYSIS_FAILED": "分析未完成，请重试",
    "SAVE_CANCELLED": "已取消保存",
    "SAVE_INPUT_OVERWRITE_FORBIDDEN": "不能覆盖当前打开的输入结果文件",
    "INTERNAL_ERROR": "程序处理请求时发生内部错误",
}


@dataclass(frozen=True)
class ApplicationErrorDTO:
    code: str
    message: str


@dataclass(frozen=True)
class AppStateDTO:
    frontend_ready: bool
    case_loaded: bool
    loading: bool
    mode: str = "local"
    api_version: str = "1"
    frontend_version: str = "0.2.0"
    schema_versions_supported: list[str] = field(default_factory=lambda: ["1.16"])
    renderer: str = "edgechromium"
    capabilities: list[str] = field(default_factory=lambda: [
        "load_standard_result", "review_modules", "paged_items",
        "evidence_inspector", "source_review", "theme", "case_switch",
        "case_analysis", "analysis_progress", "analysis_cancellation", "save_result",
    ])
    case_session_id: str | None = None
    case_revision: int = 0


@dataclass(frozen=True)
class SourceReviewItemDTO:
    source_id: str
    display_name: str
    source_type: str
    status: str
    review_reason: str
    parser_name: str | None
    generated_transactions: bool


@dataclass(frozen=True)
class SourceReviewSummaryDTO:
    case_session_id: str
    total: int
    items: list[SourceReviewItemDTO] = field(default_factory=list)


@dataclass(frozen=True)
class SourceReviewDTO:
    """Compatibility DTO retained for the validated 12B-0 slice."""
    source_name: str
    reason: str


@dataclass(frozen=True)
class CaseHeaderDTO:
    case_name: str
    period_start: str
    period_end: str
    source_count: int
    transaction_count: int
    analysis_status: str
    evidence_status: str
    schema_version: str
    case_session_id: str = ""
    case_revision: int = 0
    review_source_count: int = 0
    review_sources: list[SourceReviewDTO] = field(default_factory=list)


@dataclass(frozen=True)
class FilterOptionDTO:
    value: str
    label: str


@dataclass(frozen=True)
class FilterDefinitionDTO:
    key: str
    label: str
    kind: str
    options: list[FilterOptionDTO] = field(default_factory=list)


@dataclass(frozen=True)
class ModuleDescriptorDTO:
    module_id: str
    title: str
    icon: str
    availability: str
    display_kind: str
    total_count: int
    review_count: int
    status: str
    description: str
    supported_filters: list[FilterDefinitionDTO]
    evidence_supported: bool


@dataclass(frozen=True)
class ModuleRegistryDTO:
    case_session_id: str
    modules: list[ModuleDescriptorDTO]


@dataclass(frozen=True)
class ModuleSummaryDTO:
    module_id: str
    title: str
    total_count: int
    review_count: int
    status: str
    description: str
    boundary_note: str
    category_counts: dict[str, int]
    source_count: int
    case_session_id: str


@dataclass(frozen=True)
class ReviewItemDTO:
    item_id: str
    transaction_id: str | None
    date: str | None
    direction: str | None
    amount: str | None
    primary_text: str
    secondary_text: str | None
    counterparty: str | None
    matched_text: str | None
    interpretation: str | None
    category: str | None
    review_status: str | None
    source_name: str | None
    evidence_available: bool


@dataclass(frozen=True)
class PagedModuleItemsDTO:
    module_id: str
    case_session_id: str
    page: int
    page_size: int
    total: int
    total_pages: int
    items: list[ReviewItemDTO]
    available_filters: list[FilterDefinitionDTO]
    meta: dict[str, object] = field(default_factory=dict)


@dataclass(frozen=True)
class PurchaseSummaryDTO:
    total_count: int
    direct_count: int
    deposit_count: int
    prior_income_count: int
    review_count: int
    category_counts: dict[str, int]
    boundary_note: str


@dataclass(frozen=True)
class TransactionListItemDTO:
    transaction_id: str
    date: str
    direction: str
    amount: str
    counterparty: str
    matched_text: str
    interpretation: str
    source_name: str
    category: str
    review_status: str


@dataclass(frozen=True)
class PagedTransactionsDTO:
    items: list[TransactionListItemDTO]
    page: int
    page_size: int
    total: int
    total_pages: int
    filters: dict[str, str] = field(default_factory=dict)
    query_elapsed_ms: float = 0.0
    payload_bytes: int = 0


@dataclass(frozen=True)
class EvidenceDetailDTO:
    transaction_id: str
    transaction_id_short: str
    date: str
    direction: str
    amount: str
    counterparty: str
    summary: str
    purpose: str
    source_name: str
    page_no: int
    row_no: int
    evidence_locator: str
    reference_reason: str
    integrity_status: str
    masked_original_fields: list[str]
    full_original_fields: list[str]
    case_session_id: str = ""


def to_dict(value: Any) -> Any:
    return asdict(value) if hasattr(value, "__dataclass_fields__") else value


class ApplicationError(RuntimeError):
    def __init__(self, code: str, message: str | None = None) -> None:
        self.code = code
        super().__init__(message or ERROR_MESSAGES.get(code, ERROR_MESSAGES["INTERNAL_ERROR"]))

File: web_frontend/src/bridge/contracts.ts
export type ApiEnvelope<T> = {
  ok: boolean;
  data: T | null;
  error: { code: string; message: string } | null;
  meta: { request_id?: string; elapsed_ms: number; payload_bytes?: number };
};

export type AppStateDTO = {
  frontend_ready: boolean;
  case_loaded: boolean;
  loading: boolean;
  mode: string;
  api_version: string;
  frontend_version: string;
  schema_versions_supported: string[];
  renderer: string;
  capabilities: string[];
  case_session_id: string | null;
  case_revision: number;
};

export type SourceReviewDTO = { source_name: string; reason: string };
export type SourceReviewItemDTO = { source_id: string; display_name: string; source_type: string; status: string; review_reason: string; parser_name: string | null; generated_transactions: boolean };
export type SourceReviewSummaryDTO = { case_session_id: string; total: number; items: SourceReviewItemDTO[] };
export type CaseHeaderDTO = { case_name: string; period_start: string; period_end: string; source_count: number; transaction_count: number; analysis_status: string; evidence_status: string; schema_version: string; case_session_id: string; case_revision: number; review_source_count: number; review_sources: SourceReviewDTO[] };
export type FilterOptionDTO = { value: string; label: string };
export type FilterDefinitionDTO = { key: string; label: string; kind: "select" | "text" | "date"; options: FilterOptionDTO[] };
export type ModuleAvailability = "available" | "empty" | "not_implemented" | "unavailable";
export type ModuleDescriptorDTO = { module_id: string; title: string; icon: string; availability: ModuleAvailability; display_kind: "transaction_list" | "summary" | "disabled"; total_count: number; review_count: number; status: string; description: string; supported_filters: FilterDefinitionDTO[]; evidence_supported: boolean };
export type ModuleRegistryDTO = { case_session_id: string; modules: ModuleDescriptorDTO[] };
export type ModuleSummaryDTO = { module_id: string; title: string; total_count: number; review_count: number; status: string; description: string; boundary_note: string; category_counts: Record<string, number>; source_count: number; case_session_id: string };
export type ReviewItemDTO = { item_id: string; transaction_id: string | null; date: string | null; direction: string | null; amount: string | null; primary_text: string; secondary_text: string | null; counterparty: string | null; matched_text: string | null; interpretation: string | null; category: string | null; review_status: string | null; source_name: string | null; evidence_available: boolean };
export type PagedModuleItemsDTO = { module_id: string; case_session_id: string; page: number; page_size: number; total: number; total_pages: number; items: ReviewItemDTO[]; available_filters: FilterDefinitionDTO[]; meta: { query_elapsed_ms?: number; sort?: string } };
export type EvidenceDetailDTO = { transaction_id: string; transaction_id_short: string; date: string; direction: string; amount: string; counterparty: string; summary: string; purpose: string; source_name: string; page_no: number; row_no: number; evidence_locator: string; reference_reason: string; integrity_status: string; masked_original_fields: string[]; full_original_fields: string[]; case_session_id: string };
export type CaseSelectionDTO = { case_handle: string; case_display_name: string };
export type PreflightSourceDTO = { source_ref: string; display_name: string; extension: string; detected_source_type: string; detected_bank_type: string; supported: boolean; initial_status: string; warning: string; size: number; may_use_generic_fallback: boolean };
export type CasePreflightDTO = { case_handle: string; case_display_name: string; source_count: number; supported_source_count: number; unsupported_source_count: number; sources: PreflightSourceDTO[]; warnings: string[]; can_start: boolean; elapsed_ms: number };
export type AnalysisSourceStatusDTO = { source_ref: string; display_name: string; source_type: string; status: string; transaction_count: number; message: string };
export type AnalysisState = "running" | "cancelling" | "completed" | "failed" | "cancelled";
export type AnalysisStatusDTO = { analysis_task_id: string; state: AnalysisState; case_display_name: string; current_stage: string; current_source_name: string; completed_sources: number; total_sources: number; success_sources: number; review_sources: number; failed_sources: number; warning_count: number; started_at: string; elapsed_ms: number; cancellation_requested: boolean; error_code: string | null; error_message: string | null; result_ready: boolean; sources: AnalysisSourceStatusDTO[]; case_session_id: string | null; case_revision: number | null; transaction_count: number; result_build_ms: number | null; result_bind_ms: number | null; diagnostic_id: string | null };
export type SaveResultDTO = { saved: boolean; display_name: string };
export type RecentCaseDTO = { record_id: string; case_name: string; updated_at: string; period_start: string; period_end: string; source_count: number; transaction_count: number; analysis_status: string; schema_version: string; available: boolean };
export type RecentCasesDTO = { cases: RecentCaseDTO[]; corrupt_index: boolean };
export type ManualContextDTO = { case_name: string; saved: boolean; has_file: boolean; company_name: string; declared_work_description: string; declared_work_status: string; work_units: string[]; work_locations: string[]; residence_locations: string[]; source_names: string[]; confirmed_primary_business: string; confirmed_products_or_services: string; confirmation_note: string; confirmation_status: string; enable_ai_business_analysis: boolean };
export type ManualContextSaveDTO = { saved: boolean; case_name: string; confirmation_status: string };
export type ExportReportDTO = { saved: boolean; display_name: string };
export type ManualContextInput = { company_name: string; confirmed_primary_business: string; confirmed_products_or_services: string; confirmation_note: string; confirmation_status: string };

export interface DesktopBridge {
  getAppState(): Promise<ApiEnvelope<AppStateDTO>>;
  selectStandardResult(): Promise<ApiEnvelope<CaseHeaderDTO>>;
  chooseCaseDirectory(): Promise<ApiEnvelope<CaseSelectionDTO>>;
  inspectCaseDirectory(caseHandle: string): Promise<ApiEnvelope<CasePreflightDTO>>;
  startCaseAnalysis(caseHandle: string, options: Record<string, never>): Promise<ApiEnvelope<AnalysisStatusDTO>>;
  getAnalysisStatus(taskId: string): Promise<ApiEnvelope<AnalysisStatusDTO>>;
  cancelAnalysis(taskId: string): Promise<ApiEnvelope<AnalysisStatusDTO>>;
  dismissAnalysisTask(taskId: string): Promise<ApiEnvelope<null>>;
  saveCurrentStandardResult(): Promise<ApiEnvelope<SaveResultDTO>>;
  loadStandardResult(path: string): Promise<ApiEnvelope<CaseHeaderDTO>>;
  getCaseHeader(): Promise<ApiEnvelope<CaseHeaderDTO>>;
  getReviewModules(caseSessionId: string): Promise<ApiEnvelope<ModuleRegistryDTO>>;
  getModuleSummary(moduleId: string, caseSessionId: string): Promise<ApiEnvelope<ModuleSummaryDTO>>;
  listModuleItems(moduleId: string, page: number, pageSize: number, filters: Record<string, string>, sort: string, caseSessionId: string): Promise<ApiEnvelope<PagedModuleItemsDTO>>;
  listSourceReviews(caseSessionId: string): Promise<ApiEnvelope<SourceReviewSummaryDTO>>;
  getEvidence(transactionId: string, caseSessionId: string): Promise<ApiEnvelope<EvidenceDetailDTO>>;
  closeCase(): Promise<ApiEnvelope<null>>;
  listRecentCases(): Promise<ApiEnvelope<RecentCasesDTO>>;
  openRecentCase(recordId: string): Promise<ApiEnvelope<CaseHeaderDTO>>;
  removeRecentCase(recordId: string): Promise<ApiEnvelope<{ removed: boolean }>>;
  getManualCaseContext(caseHandle: string): Promise<ApiEnvelope<ManualContextDTO>>;
  saveManualCaseContext(caseHandle: string, fields: ManualContextInput): Promise<ApiEnvelope<ManualContextSaveDTO>>;
  rebuildContextObservations(): Promise<ApiEnvelope<CaseHeaderDTO>>;
  exportReport(): Promise<ApiEnvelope<ExportReportDTO>>;
}
